pkg/apis/cluster/v1alpha1/common_types.go

```diff
@@ -39,5 +42,25 @@ type ProviderConfig struct {
 // ProviderConfigSource represents a source for the provider-specific
 // resource configuration.
 type ProviderConfigSource struct {
-	// TODO(roberthbailey): Fill these in later
+	// The machine class from which the provider config should be sourced.
+	// +optional
+	MachineClass *MachineClassRef `json:machineClass,omitempty`
+}
+
+// MachineClassRef is a reference to the MachineClass object. Controllers should find the right MachineClass using this reference.
+type MachineClassRef struct {
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Parameters allow basic substitution to be applied to
+	// a MachineClass (where supported).
+	// Keys must not be empty. The maximum number of
+	// parameters is 512, with a cumulative max size of 256K.
+	// TODO(roberthbailey): Should this be a json-patch?
+	// +optional
+	Parameters map[string]string `json:parameters,omitempty`
```

Comment:
krousey: What kinds of things is this meant to override? Capacity? Allocatable? Or stuff in the provider config? All of these seem like they could be deeply nested structures that a map would not be able to fully express.

If ProviderConfig, perhaps we consider json patches?

roberthbailey: My thinking was in the provider config (e.g. parameterize the zone of a class to stamp out identical machine sets in multiple zones).

Reply:
I think we should leave this out for now. If we find a need for it, we can figure out how to add it later (templating vs. overlays will be contentious). But until there is a concrete need, less is more.